File: python/credit/credit.py
# get the get_int function from cs50
from cs50 import get_int

# initialize variables
credit = sum_luhn = 0
mod = 10
div = 1

# keep prompting until user gives credit card number greater than 0
while credit < 1:
    credit = get_int("Number: ")

# keeps track of how many numbers iterated over
count = 0

# get the length of the credit card
length = len(str(credit))

# apply luhn's algorithm
#pa
while count < length:
    # get the current number
    current = (credit % mod) // div

    # check to double number or not
    # before adding to sum
    if count % 2:
        current *= 2
        if current > 9: # add first and last digits seperately
            sum_luhn += (current % 10) + (current // 10)
        else:
            sum_luhn += current
    # add number to sum without doubling it
    else:
        sum_luhn += current
    mod *= 10
    div *= 10
    count += 1
# end of luhn algorightm

# check if credit card is valid
credit = str(credit) # convert the credit card to string
if  sum_luhn % 10 == 0:
    # check if visa and correct length
    if (length == 13 or length == 16) and credit[0] == "4": # validate length
        print("VISA")

    # check if valid mastercard and correct length
    elif length == 16 and (int(credit[0:2]) >= 51 and int(credit[0:2]) <= 55):
        print("MASTERCARD")

    # check if amex and correct length
    elif length == 15 and (credit[0:2] == "34" or credit[0:2] == "37"):
        print("AMEX")

    # passed the luhn algorith but not recognised
    else:
        print("INVALID")

# faiiled the mod 10 check
else:
    print("INVALID")

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int checkVisa(long credit);
int checkAmericanExpress(long credit);
int checkMasterCard(long credit);
int getLength(long credit);

int main(void)
{
    long credit = get_long("Number: ");
    int doubles = 0, skipped = 0;
    long mod = 10, div = 1;

    int count = 0, current;

    // luhn algorithm to verify credit card
    int length = getLength(credit);
    while (count < length)
    {
        current = (credit % mod) / div;
        if (count % 2)
        {
            current *= 2;
            if (current > 9)
            {
                doubles += current % 10;
                doubles += current / 10;
            }
            else
            {
                doubles += current;
            }
        }
        else
        {
            skipped += current;
        }
        mod *= 10, div *= 10;
        count++;
    }

    // check if valid credit card
    if ((doubles + skipped) % 10 == 0)
    {
        // check type of credit card
        if (checkVisa(credit))
            printf("VISA\n");
        else if (checkAmericanExpress(credit))
            printf("AMEX\n");
        else if (checkMasterCard(credit))
            printf("MASTERCARD\n");
        // credit card passed the luhn algorithm but dont know which type
        else
            printf("INVALID\n");
    }
    // invalid credit card
    else
    {
        printf("INVALID\n");
    }
}

// method to check if credit card is visa
int checkVisa(long credit)
{
    int length = getLength(credit);
    while (credit > 10)
    {
        credit /= 10;
    }
    return credit == 4 && (length == 13 || length == 16);
}

// method to check if credit card is mastercard
int checkMasterCard(long credit)
{
    int length = getLength(credit);
    while (credit > 100)
    {
        credit /= 10;
    }
    return credit >= 51 && credit <= 55 && length == 16;
}

// method to check if credit card is american express
int checkAmericanExpress(long credit)
{
    int length = getLength(credit);
    while (credit > 100)
    {
        credit /= 10;
    }
    return (credit == 34 || credit == 37) && length == 15;
}

// get the length of digits in credit
int getLength(long credit)
{
    int count = 0;
    while (credit != 0)
    {
        credit /= 10;
        count++;
    }
    return count;
}
